azkaban-exec-server/src/test/java/azkaban/dag/DagBuilderTest.java

```diff
@@ -79,6 +79,27 @@ private NodeBuilder createNodeBuilder(final String name) {
     return this.dagBuilder.createNode(name, mock(NodeProcessor.class));
   }
 
+  @Test
+  public void build_should_throw_exception_when_circular_dependency_is_detected() {
+    // given
```

Comment:
Maybe add, to help reader:
`// circular dependency like this: nb1 <- nb2 <- nb3 <- nb1`

Reply:
Done.